File: web_interface/jvox_ai/jvox_ai_magics/jvox_ai_settings_comm.py
"""
JVox AI Settings Comm Extension

An IPython extension that receives AI backend settings (ai_client, api_key)
from the JupyterLab frontend via a Comm channel, so the %%jvoxAI cell magic
can use the user-chosen AI backend.

Usage:
    Loaded automatically by the JupyterLab frontend via:
        %load_ext jvox_ai_magics.jvox_ai_settings_comm
"""

import sys

# Current AI settings, shared with jvox_ai_magics
_ai_client = "ollama"
_api_key = ""

# The Comm object for receiving settings from the frontend
_jvox_ai_settings_comm = None


def get_ai_client():
    """Return the current AI client name ('ollama' or 'gemini')."""
    return _ai_client


def get_api_key():
    """Return the current API key (may be empty)."""
    return _api_key


def _jvox_ai_settings_comm_msg(comm, open_msg):
    """Called when the frontend opens the 'jvox_ai_settings' Comm channel."""
    global _jvox_ai_settings_comm
    _jvox_ai_settings_comm = comm

    # Process any initial data sent with the open message
    if open_msg.get("content", {}).get("data"):
        _apply_settings(open_msg["content"]["data"])

    # Handle subsequent messages
    @comm.on_msg
    def _on_msg(msg):
        _apply_settings(msg["content"]["data"])

    @comm.on_close
    def _on_close(msg):
        global _jvox_ai_settings_comm
        _jvox_ai_settings_comm = None


def _apply_settings(data):
    """Apply AI settings received from the frontend."""
    global _ai_client, _api_key

    if "ai_client" in data:
        _ai_client = data["ai_client"]
    if "api_key" in data:
        _api_key = data["api_key"]

    print(
        f"JVox AI settings updated – client: {_ai_client}, "
        f"api_key: {'(set)' if _api_key else '(empty)'}",
        file=sys.stderr,
    )


def load_ipython_extension(ipython):
    """Register the Comm target when loaded via %load_ext."""
    ipython.kernel.comm_manager.register_target(
        "jvox_ai_settings", _jvox_ai_settings_comm_msg
    )


def unload_ipython_extension(ipython):
    """Unregister the extension."""
    global _jvox_ai_settings_comm
    if _jvox_ai_settings_comm is not None:
        try:
            _jvox_ai_settings_comm.close()
        except Exception:
            pass
        _jvox_ai_settings_comm = None

File: web_interface/jvox_jlab_ext/src/jvox_ai_settings_comm.ts
/**
 * JVox AI Settings Comm
 *
 * Opens a Comm channel to the kernel's jvox_ai_settings_comm IPython
 * extension so the frontend can push AI backend settings (ai_client,
 * api_key) into the kernel for use by the %%jvoxAI cell magic.
 *
 * The kernel-side extension is loaded automatically:
 *   %load_ext jvox_ai_magics.jvox_ai_settings_comm
 */

import { INotebookTracker } from '@jupyterlab/notebook';
import { IKernelConnection } from '@jupyterlab/services/lib/kernel/kernel';
import { JVoxSettings } from './jvox_settings';

export class JVoxAiSettingsComm {
    private _comm: any = null;
    private _extensionLoaded: boolean = false;
    private _settings: JVoxSettings;

    constructor(settings: JVoxSettings) {
        this._settings = settings;
    }

    /**
     * Load the IPython extension in the kernel and open the Comm channel.
     * Should be called when a kernel becomes available.
     */
    public async setup(tracker: INotebookTracker): Promise<void> {
        const kernel = tracker.currentWidget?.context?.sessionContext?.session?.kernel;
        if (!kernel) {
            console.debug('JVox AiSettingsComm: No kernel available yet.');
            return;
        }

        // Avoid re-setup if already connected to this kernel
        if (this._comm) {
            console.debug('JVox AiSettingsComm: Comm already open.');
            return;
        }

        await this._loadExtensionAndOpenComm(kernel);
    }

    /**
     * Send the current AI settings to the kernel.
     * Can be called any time after setup() succeeds.
     */
    public sendSettings(): void {
        if (!this._comm) {
            console.debug('JVox AiSettingsComm: Comm not open, cannot send settings.');
            return;
        }

        const data = {
            ai_client: this._settings.aiClient,
            api_key: this._settings.geminiApiKey,
        };

        try {
            this._comm.send(data);
            console.log('JVox AiSettingsComm: Sent settings to kernel:', {
                ai_client: data.ai_client,
                api_key: data.api_key ? '(set)' : '(empty)',
            });
        } catch (e) {
            console.error('JVox AiSettingsComm: Failed to send settings.', e);
        }
    }

    private async _loadExtensionAndOpenComm(kernel: IKernelConnection): Promise<void> {
        // Load the IPython extension in the kernel if not yet loaded
        if (!this._extensionLoaded) {
            try {
                const future = kernel.requestExecute({
                    code: '%load_ext jvox_ai_magics.jvox_ai_settings_comm',
                    silent: true,
                    store_history: false,
                });
                await future.done;
                this._extensionLoaded = true;
                console.log('JVox AiSettingsComm: IPython extension loaded.');
            } catch (e) {
                console.error('JVox AiSettingsComm: Failed to load IPython extension.', e);
                return;
            }
        }

        // Open the Comm channel and send initial settings
        try {
            this._comm = kernel.createComm('jvox_ai_settings');

            this._comm.onClose = () => {
                console.log('JVox AiSettingsComm: Comm closed.');
                this._comm = null;
            };

            // Open with the current settings as initial payload
            await this._comm.open({
                ai_client: this._settings.aiClient,
                api_key: this._settings.geminiApiKey,
            }).done;
            console.log('JVox AiSettingsComm: Comm channel opened with initial settings.');
        } catch (e) {
            console.error('JVox AiSettingsComm: Failed to open Comm.', e);
            this._comm = null;
        }
    }

    /**
     * Close the Comm channel and reset state.
     */
    public dispose(): void {
        if (this._comm) {
            try {
                this._comm.close();
            } catch (e) {
                // Ignore errors during close
            }
            this._comm = null;
        }
        this._extensionLoaded = false;
    }
}
